main.go

```diff
@@ -132,3 +131,17 @@ func movePostfixOptions(args []string) []string {
 	// append extracted arguments to the real args
 	return append(args, the_args...)
 }
+
+func parsePath(path string) (ipath.Path, error) {
+	ipfsPath, err := ipath.ParsePath(path)
+	if err == nil { // valid canonical path
+		return ipfsPath, nil
+	}
+	// allow HTTP gateway URLs as input as well
+	matches := regexp.MustCompile(`^https?://.*(/(?:ipfs|ipns|ipld)/.+)$`).FindStringSubmatch(path)
```

Comment:
`*` is greedy. This'll parse `http://bla.com/ipfs/QmA/ipfs/foobar` as `/ipfs/foobar`. What if we just parsed the input as a URL and took the path? IMO, it's reasonable to assume that gateways are run on their own origin (for security reasons).